File: app/src/main/java/sakura/felica/felicahistory/NanacoHistory.java
package sakura.felica.felicahistory;

/**
 * nanaco履歴コード
 * 参考：http://jennychan.web.fc2.com/format/nanaco.html
 */

public class NanacoHistory {
    public int remain;

    public NanacoHistory() {
    }

    public static NanacoHistory parse(byte[] res, int off) {
        NanacoHistory self = new NanacoHistory();
        self.init(res, off);
        return self;
    }

    private void init(byte[] res, int off) {
        this.remain = toInt(res, off, 3, 2, 1, 0); //12-15: Edy残高
    }

    private int toInt(byte[] res, int off, int... idx) {
        int num = 0;
        for (int i = 0; i < idx.length; i++) {
            num = num << 8;
            num += ((int) res[off + idx[i]]) & 0x0ff;
        }
        return num;
    }

    public String toString() {
        String str = "残高：" + remain + "円";
        return str;
    }
}

File: app/src/main/kotlin/sakura/felica/felicahistory/NanacoHistory.kt
package sakura.felica.felicahistory

/**
 * nanaco履歴コード
 * 参考：http://jennychan.web.fc2.com/format/nanaco.html
 */

class NanacoHistory {
  var remain: Int = 0

  private fun init(res: ByteArray, off: Int) {
    this.remain = toInt(res, off, 3, 2, 1, 0) //12-15: Edy残高
  }

  private fun toInt(res: ByteArray, off: Int, vararg idx: Int): Int {
    var num = 0
    for (i in idx.indices) {
      num = num shl 8
      num += res[off + idx[i]].toInt() and 0x0ff
    }
    return num
  }

  override fun toString(): String = "残高：" + remain + "円"

  companion object {

    fun parse(res: ByteArray, off: Int): NanacoHistory {
      val self = NanacoHistory()
      self.init(res, off)
      return self
    }
  }
}
